### PR TITLE
Bayes backprop

### PR DESCRIPTION
**Pull request by aidan-plenert-macdonald**
Showing an implementation of https://arxiv.org/pdf/1505.05424.pdf using hybrid layers

### CODE REVIEW

**Comment by zackchase** [comment]
@steverab  please review

**Comment by zackchase** [comment]
Hey  @steverab would you like to take a look or should I field it?

**Comment by steverab** [comment]
@aidan-plenert-macdonald Thanks a lot for your contribution! We were still missing a regression example trained using Bayes by Backprop, so this is a great extension. Also, the current `gluon` notebook of Bayes by Backprop does not define a new layer type, but trains the variational parameters without interfering with the networks architecture directly and can hence be thought of as a training method instead. Adding a separate layer to do this offers another way to look at this problem.

I wonder though whether this notebook should be split into two notebooks, as the two main sections are not really related to each other. I'd see great potential for a regression notebook and integrating the second part of this notebook into my `gluon` notebook to contrast the two modeling perspectives. What do you think @zackchase? Also, this notebook seems to be added to Chapter 3. Should this be moved to Chapter 18 instead?

At some point, we might also want make sure to sync the language and story line between the notebooks talking about Bayes by Backprop as there are some differences in the current notebook versions.

PS: Sorry for the late response, first week of classes at my university. ;)

**Comment by aidan-plenert-macdonald** [comment]
@steverab I totally missed Chapter 18! I'll look through it later, but it seems like you cover more in your chapter. Also, my code is missing the extra loss terms because I wasn't sure how to neatly fit it in with MxNet. Ideally, I was hoping to have the layer automatically add in the extra loss term for the log Pr(w | mu, sigma) term. Personally, I find it annoying to have to track my loss terms.

Ultimately, I was really just working to get a [BBB LSTM](https://arxiv.org/pdf/1704.02798.pdf) working. We could move the regression code where ever it works best and do whatever with the new layer, and once I get the BBB-LSTM working I'll submit a PR too.

Let me know what you'd like to do.

**Comment by steverab** [comment]
@aidan-plenert-macdonald I see. Did you take a look at Chapter 18 in the meantime? I was working on an implementation of the RNN paper, too, but was not able to fully finish and test it until the end of my internship. I passed all of this over to my mentor, Jan Gasthaus, for further development. Maybe the two of you should sync on that matter.
As I'm also still kinda new to this project, I don't feel like I'm qualified to make the final decision on how these notebook should look like. Again, I think it would be great to have a regression, a classification and a RNN-based notebook, contrast the from scratch and gluon perspectives, and model the uncertainty as part of a specific layer type or through external variables. Maybe @zackchase can also give his opinion on this matter. :)

**Comment by aidan-plenert-macdonald** [comment]
@steverab Yeah. I just started looking at it (I had to demo something else last week). My plan is to make another notebook for my regression code and then replace some of your code with my HybridBlock and modifying your loss function somehow to take advantage of my HybridBlock.

I just need to figure out how to mix a Loss function and hybrid block so that the random weight shift used in the forward pass is the same as was used in the loss computation. Granted, this probably doesn't matter too much because the randomness should average out, but I'd like to try.

**Comment by aidan-plenert-macdonald** [comment]
@steverab [Here](https://github.com/aidan-plenert-macdonald/mxnet-the-straight-dope/blob/43d8e05d4ed17ece405704a2fb18c319012583d8/chapter18_variational-methods-and-uncertainty/bayes-by-backprop-gluon.ipynb) is the preliminary change. you can see under "Model Definition" that I am trying to make the network track it's own prior loss terms.

**Comment by zackchase** [comment]
Hi @aidan-plenert-macdonald  -- I'll go through these thoroughly in next few days and weigh in properly. Thanks for the efforts! And thanks @steverab for looking the proposals over. Maybe we can chat soon outside the context of this thread - interested in your opinion on the right pedagogical flow for these sections.

**Comment by aidan-plenert-macdonald** [comment]
@steverab I have some odd training behavior that isn't present in your code. Namely, the train/test accuracy seems that it might be oscillating with the epochs. I only see an increase in the beginning then it starts to decrease after about 2-3 epochs, but I never run it long enough to see if it will increase again, but the loss in sigma is kind of like the Lennard-Jones potential, so I imagine there might be a oscillation.

I noticed you have the lines,

```python
    def log_gaussian(self, x, mu, sigma):
        return -0.5 * np.log(2.0 * np.pi) - nd.log(sigma) - (x - mu) ** 2 / (2 * sigma ** 2)
```

and I am concerned that the way you are doing it here, the whole second term might actually drop out of the compute graph. You generate your `x` via,

```python
def transform_gaussian_samples(mus, sigmas, epsilons):
    samples = []
    for j in range(len(mus)):
        samples.append(mus[j] + sigmas[j] * epsilons[j])
    return samples
```

which is tracked by `autograd.record()`, then if you assemble `(x - mu) ** 2 / (2 * sigma ** 2)`, you get,
```python
(mu + sigma*epsilon - mu) ** 2 / (2 * sigma ** 2)
```
which obviously reduces to `1/2`.  Being a constant, this term may not be contributing to the training which would explain why when I comment out my version of that line,

```python
epsilon**2 / (2 * sigma**2)
```
I get your training behavior.

Do you know what the right way to do this is?

**Comment by aidan-plenert-macdonald** [comment]
@zackchase The one problem I'd like to fix before you accept this code is odd behavior from the prior loss term. For some reason, when I include the prior loss, it trains really poorly and accuracy actually decreases over time. I probably messed up a negative sign somewhere or dropped a term accidentally.

**Comment by vishaalkapoor** [comment]
@aidan-plenert-macdonald As it's been a while since the last comment
"The one problem I'd like to fix before you accept this code ..."
and the repo has diverged quite a bit, I'll close this off, but please re-open if there's still interest and I'll take a look.

Vishaal
